FirebaseDynamicLinks/Sources/Utilities/FDLUtilities.m

```diff
@@ -204,64 +204,92 @@ BOOL FIRDLOSVersionSupported(NSString *_Nullable systemVersion, NSString *minSup
-                                     options:NSCaseInsensitiveSearch | NSAnchoredSearch]
+      if (([urlStr rangeOfString:domainURIPrefixStr
+                         options:NSCaseInsensitiveSearch | NSAnchoredSearch]
                .location) == 0) {
-        // The (short) URL needs to be longer than the domainURIPrefix, it's first character after
-        // the domainURIPrefix needs to be '/' and should be followed by at-least one more
-        // character.
-        if (urlStr.length > domainURIPrefixStr.length + 1 &&
-            ([urlStr characterAtIndex:domainURIPrefixStr.length] == '/')) {
-          // Check if there are any more '/' after the first '/'trailing the
-          // domainURIPrefix. This does not apply to unique match links copied from the clipboard.
-          // The clipboard links will have '?link=' after the domainURIPrefix.
-          NSString *urlWithoutDomainURIPrefix =
-              [urlStr substringFromIndex:domainURIPrefixStr.length + 1];
-          if ([urlWithoutDomainURIPrefix rangeOfString:@"/"].location == NSNotFound ||
-              [urlWithoutDomainURIPrefix rangeOfString:@"?link="].location != NSNotFound) {
-            customDomainMatchFound = true;
-            break;
-          }
+        NSString *urlWithoutDomainURIPrefix = [urlStr substringFromIndex:domainURIPrefixStr.length];
+        BOOL matchesRegularExpression =
+            ([urlWithoutDomainURIPrefix
+                 rangeOfString:@"((\\/[A-Za-z0-9]+)|((\\?|\\/\\?)link=https?.*))"
+                       options:NSRegularExpressionSearch]
+                 .location != NSNotFound);
+
+        if (matchesRegularExpression) {
+          customDomainMatchFound = true;
+          break;
         }
       }
     }
   }
   return customDomainMatchFound;
 }
 
+/*We are validating following domains in proper format.
+ *.page.link
+ *.app.goo.gl
+ *.page.link/i/
+ *.app.goo.gl/i/
+ */
+BOOL FIRDLIsAValidDLWithFDLDomain(NSURL *_Nullable URL) {
+  BOOL matchesRegularExpression = false;
+  NSString *urlStr = URL.absoluteString;
+
+  if ([URL.host containsString:@".page.link"] || [URL.host containsString:@".app.goo.gl"]) {
+    // Matches the *.page.link and *.app.goo.gl domains.
+    matchesRegularExpression =
+        ([urlStr rangeOfString:@"^https?://[a-zA-Z0-9]+((\\.app\\.goo\\.gl)|(\\.page\\.link))((\\/"
+                               @"?\\?link=https?.*)|(\\/[a-zA-Z0-9]+)((\\/?\\?.*=.*)?$|$))"
+                       options:NSRegularExpressionSearch]
+             .location != NSNotFound);
+
+    if (!matchesRegularExpression) {
+      // Matches the *.page.link/i/ and *.app.goo.gl/i/ domains.
+      matchesRegularExpression =
+          ([urlStr rangeOfString:
+                       @"^https?:\\/\\/[a-zA-Z0-9]+((\\.app\\.goo\\.gl)|(\\.page\\.link))\\/i\\/.*$"
+                         options:NSRegularExpressionSearch]
+               .location != NSNotFound);
+    }
+  }
```

Comment:
Is there a reason why we're not checking for `.page.link` and `.page.link/i/` together and `.app.goo.gl` with `.app.goo.gl/i/`? Seems like the regex would be simpler.

Reply:
The expected suffix for .page.link and .page.link/i are different.

https://$appdomain.page.link/i/$anything is valid but https://$appdomain.page.link/ has many restrictions to be a valid link. For example: 

https://$appdomain.page.link/p1 is valid but https://$appdomain.page.link/p1/p2 is invalid 

https://$appdomain.page.link?link=https://somedomain is valid but https://$appdomain.page.link?url=https://somedomain is invalid 

Since my expertise on regex is not that great, I find it difficult to mix these two checks together.

Reply:
Ahh, that makes sense. Adding a comment above the regex would be helpful for those that aren't familiar with the rules around the `/i/` if you don't mind

Reply:
Sure